### DISCUSSION TITLE
Question around CFD (2D) simulation dataset

### DISCUSSION THREAD

**Discussion by Forbu** [pdebench/PDEBench#20] [Q&A] [answered]
Hello everyone !

First thank you for PDEbench, It helps me in my research.

I was looking into the CFD 2D data and was trying to do a quick check on the data.

I was trying to see if the 2D CFD was respecting the continuity equation and I (think I) encounter some error : 

Here is my script to import the data to retrieve one state of the equation : (first batch, first temporal data)
```python
import numpy as np
import matplotlib.pyplot as plt
import h5py

path_hdf5 = "2D_CFD_Turb_M0.1_Eta1e-08_Zeta1e-08_periodic_512_Train.hdf5" # the file I used

idx_t = 0
idx_b = 0

delta_x = 1.0 / 512
delta_t = 5.0 / 100

with h5py.File(path_hdf5, "r") as f:

    vx = np.array(f["Vx"][idx_b, idx_t, :, :])
    vy = np.array(f["Vy"][idx_b, idx_t, :, :])

    density = np.array(f["density"][idx_b, idx_t, :, :])
    pressure = np.array(f["pressure"][idx_b, idx_t, :, :])

    vx_next = np.array(f["Vx"][idx_b, idx_t+1, :, :])
    vy_next = np.array(f["Vy"][idx_b, idx_t+1, :, :])

    density_next = np.array(f["density"][idx_b, idx_t+1, :, :])
    pressure_next = np.array(f["pressure"][idx_b, idx_t+1, :, :])
```

To check if the equation respect the continuity equation 
![image](https://user-images.githubusercontent.com/11457947/205742667-accfa121-0bec-479a-881b-9742a9862958.png)
 I used a finite difference approximation (first order) :

```python
rho_vx_div = np.zeros_like(vx * density)

v_d_x = vx * density

rho_vx_div[:, 1:-1] = (v_d_x[:, 2:] - v_d_x[:, :-2]) / (2 * delta_x) # approximation of delta rho * vx / delta x

rho_vy_div = np.zeros_like(vy)

v_d_y = vy * density

rho_vy_div[1:-1, :] = (v_d_y[2:, :] - v_d_y[:-2, :]) / (2 * delta_x)  # approximation of delta rho * vy / delta y

div = rho_vx_div + rho_vy_div # approximation of div ( rho * v)

temporal_div = (density_next - density)/ delta_t # approximation of delta rho / delta t
```

The issue is that when I check the loss I have something very different from 0 :

![image](https://user-images.githubusercontent.com/11457947/205741961-420d9ad8-b5eb-4b3d-be6f-faa2aca2f029.png)

![image](https://user-images.githubusercontent.com/11457947/205742016-a3e4d5a0-275e-42ce-bbb5-c7f7c7839828.png)

I am missing something ?

I must say that I already use this "finite difference" checking for the burger equation and the darcy equation and the dataset was OK.

**Comment by mtakamoto-D** [comment, marked as answer]
Thank you for your interest on our dataset. 
Concerning your issue, it seems a combination of (1) CFL condition and (2) spatial finite differential scheme. 
(1) when approximating d/dt part with finite differentiation method, the time-step size (D t) should be less than an upper limit described by CFL condition
https://en.wikipedia.org/wiki/Courant%E2%80%93Friedrichs%E2%80%93Lewy_condition
We subsampled temporal data, and Dt in the data is bigger than the upper limit from the CFL condition. 
(2) you seem to use central differential scheme, but we used upwind scheme (more concretely MUSCL + HLLC). 

In summary, our calculation satisfies PDE but we cannot check it using only the dataset because of the temporal sub-sampling. 
If you do want to do it, please try to perform our data-generation script and save all the data (that should be very huge datasize..). 
Then you can obtain temporal data with dt < dt_CFL.

**Reply by Forbu** [reply]
Thank you for your answer !
I will try to look more into it => it is very possible that it is the upwind scheme that screw up my verification.

**Comment by mtakamoto-D** [comment]
Hi. 
In my personal opinion, CFL condition is more important, and would be the reason of this issue. 
Thank you!
